pkg/gadgets/biolatency/tracer/core/tracer.go

```diff
@@ -0,0 +1,221 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracer
+
+// #include <linux/types.h>
+// #include "./bpf/biolatency.h"
+import "C"
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"strings"
+	"unsafe"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/link"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets"
+)
+
+//go:generate sh -c "GOOS=$(go env GOHOSTOS) GOARCH=$(go env GOHOSTARCH) go run github.com/cilium/ebpf/cmd/bpf2go -target bpfel -cc clang biolatency ./bpf/biolatency.bpf.c -- -I./bpf/ -I../../../../ -target bpf -D__TARGET_ARCH_x86"
+
+type Tracer struct {
+	node                string
+	objs                biolatencyObjects
+	blockRqCompleteLink link.Link
+	blockRqInsertLink   link.Link
+	blockRqIssueLink    link.Link
+}
+
+func NewTracer(node string) (*Tracer, error) {
+	t := &Tracer{
+		node: node,
+	}
+
+	if err := t.start(); err != nil {
+		t.Stop()
+		return nil, err
+	}
+
+	return t, nil
+}
+
+func getSlots(histMap *ebpf.Map) ([]uint32, error) {
+	key := C.struct_hist_key{}
+
+	err := histMap.NextKey(nil, unsafe.Pointer(&key))
+	if err != nil {
+		if errors.Is(err, ebpf.ErrKeyNotExist) {
+			return nil, nil
+		}
+		return nil, fmt.Errorf("error getting next key: %w", err)
+	}
+
+	hist := C.struct_hist{}
+	if err := histMap.Lookup(key, unsafe.Pointer(&hist)); err != nil {
+		return nil, err
+	}
+
+	slots := []uint32{}
+	for _, val := range hist.slots {
+		slots = append(slots, uint32(val))
+	}
+
+	return slots, nil
+}
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+// starsToString prints a line of the histogram.
+// It is a golang translation of iovsiro/bcc function print_stars as time of
+// commit:
+// 3787a0d23f75 ("libbpf-tools: improve check for fentry program support")
+func starsToString(val, valMax, width uint32) string {
+	minVal := uint32(0)
+	if val < valMax {
+		minVal = val
+	} else {
+		minVal = valMax
+	}
+
+	stars := minVal * width / valMax
+	spaces := width - stars
+
+	var sb strings.Builder
+	sb.WriteString(strings.Repeat("*", int(stars)))
+	sb.WriteString(strings.Repeat(" ", int(spaces)))
+	if val > valMax {
+		sb.WriteByte('+')
+	}
+
+	return sb.String()
+}
+
+// slotsToString convert array of values to an histogram.
+// It is a golang translation of iovisor/bcc function print_log2_hist as time of
+// commit:
+// 3787a0d23f75 ("libbpf-tools: improve check for fentry program support")
+func slotsToString(slots []uint32) string {
```

Comment:
It seems to me this is following the "standard" implementation but I think we should make it differently here. I think we should define a type structure for this gadget (as we do for other gadgets), then we should marshal to json (as done in other gadgets as well) and the client should contain this logic to "pretty print" the event. 

The standard tool already supports a -j output, I think we can take some inspiration from there to define the new type struct for this gadget. 

In summary, the changes I'm proposing are:
1. Define a data for the events generated by this gadget 
2. Store the output in json format in the trace (use -j for the standard tool and adapt the core one)
3. Move the "slotsToString" logic to the client

Reply:
Done!